src/agents/graph.py

```diff
@@ -203,6 +214,84 @@ async def _common_node(self, state: CompanionState) -> dict[str, Any]:
             SUBTASKS: state.subtasks,
         }
 
+    @staticmethod
+    def _create_gatekeeper_chain(model: IModel) -> RunnableSequence:
+        """Gatekeeper node chain to handle general queries
+        and queries that can answered from conversation history."""
+
+        prompt = ChatPromptTemplate.from_messages(
+            [
+                ("system", GATEKEEPER_PROMPT),
+                MessagesPlaceholder(variable_name="messages"),
+                ("human", "query: {query}"),
+            ]
+        )
+        return prompt | model.llm.with_structured_output(GatekeeperResponse)  # type: ignore
+
+    async def _invoke_gatekeeper_node(
+        self, state: CompanionState, user_query: str | Any
+    ) -> GatekeeperResponse | Any:
+        """Invoke the Gatekeeper node."""
+        response = await ainvoke_chain(
+            self._gatekeeper_chain,
+            {
+                "messages": state.get_messages_including_summary(),
+                "query": user_query,
+            },
+        )
+        return response
+
+    async def _gatekeeper_node(self, state: CompanionState) -> dict[str, Any]:
+        """Gatekeeper node to handle general and queries that can answered from conversation history."""
+
+        try:
+            last_human_message = next(
+                (
+                    msg
+                    for msg in reversed(state.messages)
+                    if isinstance(msg, HumanMessage)
+                ),
+            )
+            response = await self._invoke_gatekeeper_node(
+                state, last_human_message.content
+            )
+
+            if response.forward_query:
+                logger.debug("Gatekeeper node forwarding the query")
+                return {
+                    NEXT: SUPERVISOR,
+                    MESSAGES: [
+                        AIMessage(
+                            content="",
+                            name=GATEKEEPER,
+                        )
+                    ],
```

Comment:
do you really need to explicityl add messages? We don't need any message from gatekeeper.